### PR TITLE
1537 action xero accounting new actions

### PR DESCRIPTION
**Pull request by bush4u**
Danny confirmed it is okay to skip **New or updated subscription** trigger

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

**Review by jcortes — CHANGES_REQUESTED** [review]
Hi @bush4u I've just left a request change I would like you to make in the meantime, thanks for this work, it's looking great so far.

File: components/xero_accounting_api/xero_accounting_api.app.mjs [hunk 1]
```diff
@@ -1,11 +1,70 @@
...
+      });
+    },
+    async getSubscription(tenant_id, modifiedSince = null) {
+      const newQueryParam = chainQueryString(queryParam);
+      const params = newQueryParam && { Where: newQueryParam };
+      return await axios(this.$auth, {
+        url: constants.INVOICE_API,
+        headers: this.getHeader(tenant_id, modifiedSince),
+        params,
+      });
```
[the remark this answers is not in the document]
Comment (bush4u) [reply components/xero_accounting_api/xero_accounting_api.app.mjs]: Done

**Review by jcortes — CHANGES_REQUESTED** [review]
Hi @bush4u thanks for making the refactor. It's looking good so far. I've just added a few more request changes and I think it'll be good for QA!

File: components/xero_accounting_api/common/common.util.mjs [hunk 2]
```diff
@@ -0,0 +1,120 @@
+import { ConfigurationError } from "@pipedream/platform";
```
[the remark this answers is not in the document]
Comment (bush4u) [reply components/xero_accounting_api/common/common.util.mjs]: Fixed

File: components/xero_accounting_api/actions/add-line-item-to-invoice/add-line-item-to-invoice.mjs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        }\``,
+    },
+  },
+  async run({ $ }) {
+    const { tenant_id, InvoiceID, LineItems } = this;
+    const data = removeNullEntries({
+      InvoiceID,
+      LineItems: formatArrayStrings(LineItems, constant.ALLOWED_LINEITEMS_KEYS),
+    });
+    return await this.xero_accounting_api.createInvoice(tenant_id, data);
```
[the remark this answers is not in the document]
Comment (bush4u) [reply components/xero_accounting_api/actions/add-line-item-to-invoice/add-line-item-to-invoice.mjs]: Fixed

File: components/xero_accounting_api/actions/add-line-item-to-invoice/add-line-item-to-invoice.mjs [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import {
+  formatArrayStrings,
+  removeNullEntries,
+} from "../../common/common.util.mjs";
+import constant from "../../common/common.constants.mjs";
+import xero_accounting_api from "../../xero_accounting_api.app.mjs";
```
[the remark this answers is not in the document]
Comment (bush4u) [reply components/xero_accounting_api/actions/add-line-item-to-invoice/add-line-item-to-invoice.mjs]: Fixed

File: components/xero_accounting_api/sources/new-updated-contact/new-updated-contact.mjs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      },
+    },
+    db: "$.service.db",
+  },
+  dedupe: "unique",
+  async run() {
+    let lastDateChecked;
+
+    this.db.get("lastDateChecked") &&
+      (lastDateChecked = this.db.get("lastDateChecked"));
```
[the remark this answers is not in the document]
Comment (bush4u) [reply components/xero_accounting_api/sources/new-updated-contact/new-updated-contact.mjs]: Done

**Comment by bush4u** [comment]
> Hi @bush4u Unable deploy add-line-item-to-invoice action, getting following error
> 
> image

This is done

**Comment by DilanAthukorala** [comment]
@bush4u  please check the following issues;

create_bill action - **DueDate** is not passing to the xero accounting site. 

find_invoice action - following error observed for tenant id
image



new-updated-contacts / new-updated-invoice sources - observing following error when deploying
image

**Comment by bush4u** [comment]
> @bush4u please check the following issues;
> 
> create_bill action - **DueDate** is not passing to the xero accounting site.
> 
> find_invoice action - following error observed for tenant id image
> 
> new-updated-contacts / new-updated-invoice sources - observing following error when deploying image

I am reproducing the error

**Comment by bush4u** [comment]
I think that error came up because the DueDate format does not follow **YYYY-MM-DD** format

**Comment by bush4u** [comment]
> > @bush4u please check the following issues;
> > create_bill action - **DueDate** is not passing to the xero accounting site.
> > find_invoice action - following error observed for tenant id image
> > new-updated-contacts / new-updated-invoice sources - observing following error when deploying image
> 
> I am reproducing the error

I was able to trap this particular error and it is fixed

**Comment by DilanAthukorala** [comment]
@bush4u Thanks the **find_invoice** bug is fixed. Please update me once the other issues are fixed. 

https://github.com/PipedreamHQ/pipedream/pull/2744#issuecomment-1124247412

**Comment by bush4u** [comment]
Thanks @DilanAthukorala ,
The bug has been fixed. Thanks

**Comment by bush4u** [comment]
Please give me a little while. Not sure why my task is breaking at this time. I will test the sources again end to end

**Comment by bush4u** [comment]
Thanks for your patience. Most of the bug were due to previous change requested.
I have tested the sources again and they should be fine now

**Comment by bush4u** [comment]
Hi  @DilanAthukorala , Kindly let mw know if I have any more clarification is required

**Comment by DilanAthukorala** [comment]
@bush4u It's better if we can some data to the events in **'New or updated invoice'** source. 
image

**Comment by bush4u** [comment]
> 
@DilanAthukorala 
Hi. 
I have discovered and fixed the bug

**Comment by bush4u** [comment]
> @bush4u I am still seeing the issue for **create_bill** action where **duedate** is not updating in the xero accounting side. image image

@DilanAthukorala
Hi.
I have discovered and fixed the bug

**Comment by bush4u** [comment]
> @bush4u It's better if we can some data to the events in **'New or updated invoice'** source. image

Done!

**Comment by DilanAthukorala** [comment]
Verification completed successfully.

**Comment by bush4u** [comment]
Hi @michelle0927,

Kindly help review for approval. 
Ticket already passed QA, and it is ready for release

**Review by michelle0927 — APPROVED** [review]

**Pull request merged**